engine/runtime/src/main/java/org/enso/interpreter/node/callable/dispatch/LoopingCallOptimiserNode.java

```diff
@@ -54,31 +57,78 @@ public static LoopingCallOptimiserNode build() {
    * @param loopNode a cached instance of the loop node used by this node
    * @return the result of executing {@code function} using {@code arguments}
    */
-  @Specialization
-  public Object dispatch(
+  @Specialization(guards = "warnings == null")
+  public Object cachedDispatch(
       Function function,
       CallerInfo callerInfo,
       State state,
       Object[] arguments,
+      Warning[] warnings,
       @Cached(value = "createLoopNode()") LoopNode loopNode) {
+    return dispatch(function, callerInfo, state, arguments, loopNode);
+  }
+
+  @Specialization(guards = "warnings != null")
+  public Object cachedDispatchWarnings(
+      Function function,
+      CallerInfo callerInfo,
+      State state,
+      Object[] arguments,
+      Warning[] warnings,
+      @Cached(value = "createLoopNode()") LoopNode loopNode) {
+    Object result = dispatch(function, callerInfo, state, arguments, loopNode);
+    return WithWarnings.appendTo(EnsoContext.get(this), result, warnings);
```

Comment:
This looks correct. If there are warnings, remove them, handle all the tail calls and attach them again. Not sure how that can be triggered by [the test](https://github.com/enso-org/enso/pull/7116/files#r1240600638) however?

Reply:
> Not sure how that can be triggered by [the test](https://github.com/enso-org/enso/pull/7116/files#r1240600638) however?

I'm not sure how to convince you other than running it with debugger attached and putting the breakpoint yourself.